first working prototype

rccrdpccl opened the pull request:
THe purpose of this project is to trigger it post-deployment, so we can check the health of the application through prometheus.
If all is well, we will use the result to auto-promote the application

rccrdpccl committed:
first working prototype

osherdp commented on Dockerfile.build:
```diff
@@ -0,0 +1,8 @@
+COPY requirements.txt requirements-dev.txt ./
+
+RUN python3 -m pip install -I --no-cache-dir -r requirements.txt &&\
+ python3 -m pip install -I --no-cache-dir -r requirements-dev.txt
```
```suggestion
RUN python3 -m pip install -I --no-cache-dir -r requirements.txt -r requirements-dev.txt
```

osherdp commented on .dockerignore:
```diff
@@ -0,0 +1,2 @@
+./openshift/
+__pycache__
```
also ``.git`` shouldn't matter too much for it, right?

osherdp commented on skipper.yaml:
```diff
@@ -0,0 +1,5 @@
+build_container_image: build
+
+containers:
+  assisted-service-build: Dockerfile.build
```
?

rccrdpccl replied:
Copy paste error, sorry :sweat:

osherdp commented on setup.py:
```diff
@@ -0,0 +1,48 @@
+from setuptools import setup, find_packages
+
+def read_file(name):
+    with io.open(os.path.join(os.path.dirname(__file__), name), encoding='utf8') as f:
```
seems like you don't import this ``io`` module (also it doesn't seem needed, no?)

osherdp replied:
also remove the ``encoding='utf8'``, I think that's the default if doing ``open(<path>, "rt")``

osherdp commented on setup.py:
```diff
@@ -0,0 +1,48 @@
+    with io.open(os.path.join(os.path.dirname(__file__), name), encoding='utf8') as f:
+        return f.read()
+
+def read_reqs(name):
+    return [line for line in read_file(name).split('\n') if line and not line.strip().startswith('#')]
```
maybe let's have the reqs in ``setup.py`` only?
either way it's more convenient to do ``pip install .`` than anything else

rccrdpccl replied:
I would like to keep `requirements.txt` as it's easier to handle caching dependencies in the dockerfile, as far as I can see.
Also I wouldn't want to maintain them in both files.
Shall I maybe just skip them in `setup.py`? WDYT?

osherdp replied:
no it's okay. leave it as-is

osherdp commented on tests/unit/test_conditions.py:
```diff
@@ -0,0 +1,113 @@
+                "interval_seconds": 30,
+                "threshold": {
+                    "upper": 10,
+                    "lower": 2,
```
can you add a comment with the field that is missing? e.g.:
```suggestion
                    "lower": 2,
                    # notice "my_field" is missing here
```

osherdp commented on tests/unit/test_evaluator.py:
```diff
@@ -0,0 +1,142 @@
+            1662137902.2000000,  # 4th execution
+            1662137907.2000000,  # 5th execution
+            1662137912.2000000,  # 6th execution
+            1662137917.2000000,  # 7th execution
```
those specific values are required for this test?
I mean, can we replace those values with the mocker returning values like ``1,2,3,...`` and only assert values of ``mock_calls `` and ``call_count``?

rccrdpccl replied:
yes sure, we could do with sequential numbers simulating ticking. I thought I'd make it realistic but yeah it's a bit unreadable

osherdp commented on promql_tester/evaluator.py:
```diff
@@ -0,0 +1,96 @@
+        return expiration_time < execution_time
+
+
+def _is_valid_res(condition, result):
```
```suggestion
def _is_valid_result(condition, result):
```

rccrdpccl replied:
You always catch my abbreviations :sweat_smile:

osherdp commented on promql_tester/evaluator.py:
```diff
@@ -0,0 +1,96 @@
+
+
+class Evaluator:  # pylint: disable=too-few-public-methods
+    """This clas handles the logic how and when to evaluate a list of conditions."""
```
```suggestion
    """This class handles the logic how and when to evaluate a list of conditions."""
```

osherdp commented on promql_tester/condition.py:
```diff
@@ -0,0 +1,110 @@
+        self._uuid = uuid.uuid4()
+        self._threshold = self._get_threshold(raw_config)
+
+    def get_id(self) -> str:
```
doesn't it return a UUID object?

osherdp commented on openshift/template.yaml:
```diff
@@ -0,0 +1,55 @@
+    spec:
+      containers:
+      - name: promql-tester
+        image: quay.io/rh_ee_rpiccoli/promql-tester:edge
```
just so I'll know, in the future it will hold the app-sre image location?

rccrdpccl replied:
The idea was that this repo just pushes the image to a quay (possibly openshift-assisted or something, not my personal namespace).
We then would reference it (hardcoded) from the project(s) that needs this validation for autopromotion.
What do you think about this?
Also I just realized template is not needed, so I will leave only example directory for a sample deployment (also useful to test against a real prometheus instance)

osherdp commented on openshift/template.yaml:
```diff
@@ -0,0 +1,55 @@
+          value: "/etc/config/conditions.yaml"
+        command:
+        - "sleep"
+        - "infinity"
```
what are those?

osherdp commented on example/deployment.yaml:
```diff
@@ -0,0 +1,49 @@
+apiVersion: apps/v1
```
how this file is different than the template?

osherdp commented on promql_tester/condition.py:
```diff
@@ -0,0 +1,110 @@
+            raise MalformedConfigException(f"`conditions` key missing from {filename}")
+        for raw_condition in config["conditions"]:
+            conditions.append(Condition(raw_condition))
+    return conditions
```
```suggestion
    with open(filename, 'r', encoding='utf-8') as config_file:
        config = yaml.safe_load(config_file)
        if "conditions" not in config:
            raise MalformedConfigException(f"`conditions` key missing from {filename}")
        
        return [Condition(raw_condition) for raw_condition in config["conditions"]]
```

osherdp commented on promql_tester/condition.py:
```diff
@@ -0,0 +1,110 @@
+            threshold["lower"] = self.DEFAULT_THRESHOLD["lower"]
+        threshold["upper"] = float(threshold["upper"])
+        threshold["lower"] = float(threshold["lower"])
+        return threshold
```
isn't it equivalent to ?
```suggestion
        return {**self.DEFAULT_THRESHOLD, **raw_config.get("threshold", {})}
```

osherdp commented on promql_tester/evaluator.py:
```diff
@@ -0,0 +1,96 @@
+            res = self._prom.custom_query(query=expr)
+            self._logger.debug(f"Query for condition {condition.get_id()} returned: {res}")
+            if not _is_valid_res(condition, res):
+                self._logger.info(f"Failed condition: {condition}")
```
maybe warning or error?

osherdp reviewed (changes requested):
partial review

adriengentil commented on promql_tester/condition.py:
```diff
@@ -0,0 +1,110 @@
+        """Returns autogenerated condition ID"""
+        return self._uuid
+
+    def get_interval_seconds(self):
```
I would suggest to have properties instead of getters/setters
```suggestion
    @property
    def interval_seconds(self):
```

rccrdpccl replied:
I can't seem to grasp what are the advantages of this approach, is this just more pythonic or is there more to it?

osherdp replied:
For Python developers if it looks like an object property they prefer getting it as such (instead of Java let's say that prefers declaring it as part of the object's behavior)

adriengentil replied:
yes, it's more pythonic. From what I know, in python the attributes are public by default, we access to them via `my_class.my_attr`. The `@property` decorator allows us to encapsulate the attributes inside the class while keeping the same access logic.

adriengentil commented on promql_tester/__main__.py:
```diff
@@ -0,0 +1,61 @@
+    logger = get_logger()
+
+    prom_url = os.getenv('PROMETHEUS_URL')
+    conditions_path = os.getenv('CONDITIONS_PATH')
```
```suggestion
    prom_url = os.environ['PROMETHEUS_URL']
    conditions_path = os.environ['CONDITIONS_PATH']
```
They seem required, let's fail early.

adriengentil commented on promql_tester/evaluator.py:
```diff
@@ -0,0 +1,96 @@
+            return False
+        return True
+
+    def _evaluate(self, condition):
```
the function always returns `True` or raise an exception, right? In that case I think the function should not return anything

adriengentil reviewed (commented):
Did a quick review, I'll look more into it tomorrow, but it looks good :+1:

adriengentil reviewed (approved):
:rocket:

rccrdpccl commented:
just fixed a couple of details in the example template:
- image referencing not my personal namespace
- command for the application instead of sleeping
- changed syntax for conditions (in the example)
